optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/common/BavetNodeBuildPolicy.java

```diff
@@ -37,18 +37,11 @@ public BavetNodeBuildPolicy(BavetConstraintSession session, int constraintCount)
         constraintIdToScoringNodeMap = new LinkedHashMap<>(constraintCount);
     }
 
-    public void updateNodeOrderMaximum(int nodeOrder) {
-        if (nodeOrderMaximum < nodeOrder) {
-            nodeOrderMaximum = nodeOrder;
-        }
-    }
-
     public <Node_ extends BavetAbstractNode> Node_ retrieveSharedNode(Node_ node) {
         Node_ sharedNode = (Node_) sharableNodeMap.computeIfAbsent(node, k -> node);
-        if (node.getNodeOrder() != sharedNode.getNodeOrder()) {
-            throw new IllegalStateException("Impossible state: the node (" + node
-                    + ")'s nodeOrder (" + node.getNodeOrder() + ") differs from the sharedNode (" + sharedNode
-                    + ")'s nodeOrder (" + sharedNode.getNodeOrder() + ").");
```

Comment:
All node indexes are unique now. Therefore this fail-fast makes no sense anymore.

Reply:
Crap, ideally when a node is replaced by a sharedNode, it doesn't take up an extra nodeIndex.

Because it does, the nodeSize will be actually bigger than the number of actual nodes in place. (so it shouldn't be called nodeSize indeed and a comment should specify that strange behaviour). So the queue will have a number of empty spaces. Actually, that hurts performance...
... so we need to recycle nodeIndexes of nodes that get shared. Maybe upon sharing, call BuildPolicy.undoNextIndex() that does nextNodeIndex--?

Reply:
This is not an issue. Node indexes are perfectly sequential, they are never duplicated. (The "+1 trick" takes care of that.)